qt/aqt/log.py

```diff
@@ -0,0 +1,102 @@
+# Copyright: Ankitects Pty Ltd and contributors
+# License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+# original from: https://github.com/abdnh/ankiutils/blob/master/src/ankiutils/log.py
+# To use inside an addon:
+#     from aqt import mw
+#     log = mw.addonManager.getLogger(__name__)
+
+
+from __future__ import annotations
+import sys
+from typing import Any
+from pathlib import Path
+import logging
+import logging.handlers
+
+
+class RotatingFileHandler(logging.handlers.RotatingFileHandler):
+    MAXSIZE = 3 * 1024 * 1024
+    BACKUPCOUNT = 5
+    def __init__(self, filename, *args, **kwargs):
+        super().__init__(filename, "a", encoding="utf-8", maxBytes=self.MAXSIZE, backupCount=self.BACKUPCOUNT)
+
+    def _on_close(self, _: Any, m: str, *args: Any, **kwargs: Any) -> None:
+        breakpoint()
+        self.close()
+
+
+class LoggerManager(logging.Manager):
+    TAG = "addon."   # all logs with name addons. will have their content saved to a separate file log
+
+    def __init__(self, path: Path | str, rootnode: logging.Logger):
+        super().__init__(rootnode)
+        self.path = Path(path)
+
+    # fixme: this seems not called
+    def _cleanup(self, handler: RotatingFileHandler):
+        from anki.hooks import wrap
+        from aqt.addons import AddonManager
+        AddonManager.deleteAddon = wrap(  # type: ignore[method-assign]
```

Comment:
It's more clean/correct to call the cleanup method inside the relevant AddonManager methods instead of patching them.

Reply:
Yes.

Reply:
Updated in [here](https://github.com/ankitects/anki/pull/2969/commits/62abefe33abed743f652fd8c4ba24f35c2a32188).